content/blog/ebpf-portability/index.md

```diff
@@ -0,0 +1,189 @@
+
+If you’re fortunate enough to work in a uniform environment, this may not be such a big deal. But if you’re writing eBPF programs for general distribution, then you don’t get to control the environment. Your users will have a variety of Linux distributions, with different kernel versions, kernel configurations, and distribution-specific quirks.
+
+Faced with such a problem, what can you do to make sure that your eBPF-based application will work on as many environments as possible?
+
+In this blog post, we examine this question, and share some of our learnings from deploying Pixie across a wide range of environments.
+
+## What's the Problem?
+
+_Note: The problem of BPF portability is covered in detail by Andrii Nakryiko in his [blog post](https://nakryiko.com/posts/bpf-portability-and-co-re/) on the subject. In this section, we rehash the problem briefly._
+
+To understand why it can be problematic to deploy eBPF programs across different target environments, it’s important to first review the eBPF build pipeline. We’ll start with the basic flow used by frameworks like BCC. There are newer approaches with libbpf + CO-RE, but we’ll cover that later.
+
+::: div image-xl
+<svg title='The BCC eBPF deployment flow: The eBPF code is compiled on the target host to make sure that the program is compatible.' src='bcc-ebpf-diagram.png' />
+:::
+
+In the basic flow, the eBPF code is compiled into BPF byte code, and then deployed into the kernel. Assuming that the BPF verifier doesn’t reject the code, the eBPF program is then run by the kernel whenever triggered by the appropriate event.
+
+In this flow, it’s important to note that this entire process needs to happen on the host machine. One can’t simply compile to BPF bytecode on their local machine and then ship the bytecode to different host machines.
+
+Why? Because each host may have a different kernel, and so kernel struct layouts may have changed.
+
+Let’s make this more concrete with a couple of examples. First, let’s look at a very simple eBPF program that doesn’t have any portability issues:
+
+```cpp
+// Map that stores counts of recv calls by PID
+BPF_HASH(counts_by_pid, uint32_t, int64_t);
+
+// Probe that counts every time it is triggered.
+// Can be used to count things like syscalls or particular functions.
+int syscall__probe_counter(struct pt_regs* ctx) {
+  bpf_trace_printk("triggered");
+  uint32_t tgid = bpf_get_current_pid_tgid() >> 32;
+
+  int64_t kInitVal = 0;
+  int64_t* count = counts_by_pid.lookup_or_init(&tgid, &kInitVal);
+  if (count != NULL) {
+	*count = *count + 1;
+  }
+
+  return 0;
+}
+```
+
+The code above is for a kprobe on the `recv()` syscall. Every time a `recv()` call is made and the probe is triggered, a count is incremented. The count is stored in a BPF map, which also means the current count can be read from user-space at any time to get the latest value.
```

Comment:
I had made changes to the snippet, but forgot to update the text. Can you please change to:

The code above can be attached on a syscall (for example, the `recv()` syscall). Then, every time the syscall is made, the probe is triggered and the count for that PID is incremented. The counts are stored in a BPF map, which means the current counts for each PID can be read from user-space at any time to get the latest value.